src/app/frontend/common/services/global/history.ts

```diff
@@ -0,0 +1,47 @@
+// Copyright 2017 The Kubernetes Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {Injectable} from '@angular/core';
+import {RawParams, StateService, TransitionPromise, TransitionService} from '@uirouter/core';
+
+@Injectable()
+export class HistoryService {
+  private previousStateName: string;
+  private previousStateParams: RawParams;
+
+  constructor(
+      private readonly state: StateService, private readonly transitions: TransitionService) {}
+
+  /** Initializes the service. Must be called before use. */
+  init(): void {
+    this.transitions.onSuccess({}, (transition) => {
+      this.previousStateName = transition.from().name || '';
+      this.previousStateParams = transition.params('from');
+    });
+  }
+
+  /**
+   * Goes back to previous state or to the provided default if none set.
+   * @param {string} defaultState
+   */
+  back(defaultState?: string): TransitionPromise {
+    let targetState = this.previousStateName || defaultState;
+
+    // If previous state is same as current state then go to default state to avoid loop.
+    if (this.state.current.name === this.previousStateName) {
```

Comment:
As you have made `defaultState` an optional parameter then you have to handle a case where previous state is same as current and `defaultState` is not set (is undefined).

Reply:
Will revert `defaultState` back to optional params. The logic behind determining the "correct" state in navigations is too ambiguous and is subject to features. I guess it's better leave this to the user of this function.